qdr/_ranker.cc

```diff
@@ -0,0 +1,416 @@
+    word_counts_t ret(document.size() * 3);
+    for (doc_t::const_iterator it = document.begin();
+        it != document.end(); ++it)
+    {
+        word_counts_t::iterator got = ret.find(*it);
+        if (got == ret.end())
+        {
+            // first time we've seen this word
+            ret[*it] = 1;
+        }
+        else
+            ret[*it] += 1;
+    }
+    return ret;
+}
+
+struct scores_t {
+    double tfidf;
+    double bm25;
+    double lm_jm;
+    double lm_dirichlet;
+    double lm_ad;
+};
+
+struct lm_scores_t {
+    double jm;
+    double dirichlet;
+    double ad;
+};
+
+class QDR
+{
+    public:
+        QDR(counts_t& counts, uint64_t total_docs);
+        ~QDR();
+
+        // compute the similarity scores
+        scores_t score(doc_t& document, doc_t& query);
+
+        // compute similiary scores for a single doc but list of queries
+        std::vector<scores_t> score_batch(doc_t& document,
+            std::vector<doc_t>& queries);
+
+        // get the IDF for a given word
+        double get_idf(const std::string& word);
+
+    private:
+        counts_t counts;
+        uint64_t total_docs;
+        uint64_t nwords_vocab;
+        uint64_t nwords;
+
+        // useful constant for language model
+        double n2p1;
+
+        // compute the TF-IDF similarity
+        double tfidf(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts);
+
+        // BM25
+        double okapi_bm25(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts,
+            const double& Ld);
+
+        // language model
+        lm_scores_t lm(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts,
+            const double& sum_count_w_given_doc);
+
+        // low level function used by public methods
+        scores_t score_single(const word_counts_t& doc_counts,
+            const word_counts_t& query_counts,
+            const double& nwords_document);
+
+        // disable some default constructors
+        QDR();
+        QDR& operator= (const QDR& other);
+        QDR(const QDR& other);
+
+};
+
+QDR::QDR(counts_t& counts_in, uint64_t total_docs)
+{
+    this->total_docs = total_docs;
+
+    // copy the counts data to private member and add up the total
+    // number of words and words in vocab
+    nwords_vocab = 0;
+    nwords = 0;
+    counts.clear();
+
+    for (counts_t::iterator it = counts_in.begin(); it != counts_in.end();
+        ++it)
+    {
+        counts[it->first] = it->second;
+        ++nwords_vocab;
+        nwords += it->second.first;
+    }
+
+    n2p1 = (double) nwords + (double) nwords_vocab + 1.0;
+}
+
+QDR::~QDR() {}
+
+double QDR::get_idf(const std::string& word)
+{
+    /// Get log(total docs / doc count for word) == IDF
+    double doc_freq;
+    counts_t::iterator got = counts.find(word);
+    if (got != counts.end())
+    {
+        // this word is in the corpus
+        doc_freq = (double) got->second.second;
```

Comment:
Is there a downside/efficiency hit to using a cast?  I sometimes like them since it makes it explicit what the calculation and has some future proofness if I come along in a few months and change the type of doc_freq but forget to add the cast back.